Add LB and URL map resources for tiles-tlog

cmurphy opened the pull request:
This change adds a load balancer with a CDN and SSL for tiles_tlog.
Since the load balancer must conditionally route to either the
Kubernetes service or a bucket, it can't be managed in Kubernetes by the
Ingress resource and instead must be managed explicitly in Terraform.
The Helm chart will use annotations on the Service to create Network
Endpoint Groups[1] to which the load balancer can route.

Since the NEGs are created by Kubernetes, they won't exist the first time these
Terraform modules are run for a given shard. The order of operations on
first deployment of any shard is:

1. Run Terraform with network_endpoint_group_zones in the tiles-tlog
   module set to empty.
2. Update the ArgoCD configuration to deploy the new tlog shard.
3. Update the Terraform input to set network_endpoint_group_zones to the
   all the zones for the GKE cluster's region.

[1] https://cloud.google.com/load-balancing/docs/negs/zonal-neg-concepts

[1] https://cloud.google.com/load-balancing/docs/negs/zonal-neg-concepts

This builds on #21, can merge this into that PR or merge them separately.

Fixes https://github.com/sigstore/rekor-tiles/issues/239
Fixes https://github.com/sigstore/rekor-tiles/issues/44
Fixes https://github.com/sigstore/rekor-tiles/issues/240
<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficient time for discussions to take place. Please use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary
<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve? How can reviewers test this PR?
-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->

#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

haydentherapper commented on gcp/modules/tiles_lb/main.tf:
```diff
@@ -0,0 +1,182 @@
+    content {
+      name            = path_matcher.key
+      default_service = google_compute_backend_service.k8s_backend_service[path_matcher.key].id
+      route_rules {
```
For when we freeze the shard, we'll need to disable this rule. Can we make this optional for a given host, maybe have a bool configuration for if the host is frozen?

cmurphy commented:
Moved this back into the tiles_tlog module, so when the module is re-instantiated for a new shard all load balancer-related resources will be duplicated for the new shard rather than being shard across shards, including the IPv4 address, A record, and SSL certificate. Also added the gRPC port and TLS certificate management here.

bobcallaway commented on gcp/modules/tiles_tlog/network.tf:
```diff
@@ -33,3 +36,221 @@ resource "google_dns_record_set" "A_tlog" {
+  ssl_policy       = google_compute_ssl_policy.ssl_policy.id
+}
+
+resource "google_compute_global_forwarding_rule" "http_forwarding_rule" {
+  count   = var.dns_domain_name == "" ? 1 : 0
+  name    = "${var.shard_name}-${var.dns_subdomain_name}-http-forwarding-rule"
+  project = var.project_id
+
+  ip_address            = google_compute_global_address.gce_lb_ipv4.address
+  target                = google_compute_target_http_proxy.lb_proxy[count.index].id
+  port_range            = "80"
+  load_balancing_scheme = "EXTERNAL_MANAGED"
+}
```
Just curious why we'd want to expose `http` endpoints at all (just `https` would be my vote)

haydentherapper replied:
Agreed, if we aren't planning a deployment with HTTP, I'd omit.

cmurphy replied:
It's tied to the DNS name. The Google-managed cert will never move past the `PROVISIONING` stage if I use a domain name I don't own. So for development purposes it was simpler to have flexibility here. I can remove all these conditionals but it means everyone who develops on this module will have to provision their own long-lived domain name.

haydentherapper replied:
Chatted offline, one option is to use one of Sigstore's test domains. This doesn't solve the problem for other developers who don't have access to Sigstore's test infra though and don't want to manage a domain.

I'm fine with leaving this in, though maybe a comment that a real-world deployment shouldn't use HTTP?

cmurphy replied:
Elaborated in the dns_domain_name variable description and expanded the comment on line 198.

haydentherapper commented on gcp/modules/tiles_tlog/network.tf:
```diff
@@ -33,3 +36,221 @@ resource "google_dns_record_set" "A_tlog" {
+  default_service = google_compute_backend_service.k8s_http_backend_service.id
+
+  host_rule {
+    hosts        = var.dns_domain_name == "" ? ["*"] : ["${var.shard_name}.${var.dns_subdomain_name}.${var.dns_domain_name}"]
```
When is `dns_domain_name` unset? Wondering if we should always expect hosts to be specified so we don't accidentally set up a URL map for all of our hosts?

cmurphy replied:
This was a pattern I introduced in https://github.com/sigstore/scaffolding/pull/1087 because I was hesitant to spend money on a domain name during development. DNS domain names last for at least a year, they can't be removed when you're done for the day and they live on even if you delete the project they belong to. If I remove these, I would have to comment out several resources during development and I would never be able to see the system working end-to-end. With this, I can at least have a pretty good idea that it's mostly working, even if I never set up a domain and can't get a valid HTTPS cert.

haydentherapper commented on gcp/modules/tiles_tlog/network.tf:
```diff
@@ -33,3 +36,221 @@ resource "google_dns_record_set" "A_tlog" {
+
+  dynamic "backend" {
+    for_each = data.google_compute_network_endpoint_group.k8s_http_neg
+    iterator = neg
```
To clarify, this is repeated because we have multiple NEGs since they're zonal, correct?

cmurphy replied:
Correct
